pkg/tuf/client.go

```diff
@@ -15,137 +15,184 @@
 		return nil, err
 	}
+	if err = c.loadMetadata(); err != nil {
+		return nil, err
+	}
 
-	tufClient := tufclient.NewClient(fileJSONStore, tufRemoteStore)
-	targetFiles, err := tufClient.Update()
+	return &c, nil
+}
+
+// DefaultClient returns a Sigstore TUF client for the public good instance
+func DefaultClient() (*Client, error) {
+	opts, err := DefaultOptions()
 	if err != nil {
 		return nil, err
 	}
+	return New(opts)
+}
 
-	// Now that we've updated, see if remote trustedroot metadata matches local disk
-	trustedrootMeta, ok := targetFiles[TrustedRootTUFPath]
-	if !ok {
-		return nil, fmt.Errorf("Unable to find %s via TUF", TrustedRootTUFPath)
+// loadMetadata controls if the client actually should perform a TUF refresh.
+// The TUF specification mandates so, but for certain Sigstore clients, it
+// may be beneficial to rely on the cache, or in air-gapped deployments it
+// it may not even be possible.
+func (c *Client) loadMetadata() error {
+	// Load the metadata into memory and verify it
+	if err := c.up.Refresh(); err != nil {
+		// this is most likely due to the lack of metadata files
+		// on disk. Perform a full update and return.
+		return c.Refresh()
 	}
 
-	trustedroot, ok := tufMetaMap[TrustedRootTUFPath]
-	if ok {
-		if ok, _ := validTarget(trustedrootMeta, trustedroot); ok {
-			return trustedroot, nil
+	var tm = c.up.GetTrustedMetadataSet()
+	if c.opts.ForceCache {
+		// Use cache until it expires
+		if tm.Timestamp.Signed.IsExpired(time.Now()) {
+			return c.Refresh()
 		}
-	}
 
-	// What's on disk didn't match, so download from TUF remote (and cache it to disk)
-	writer := &Writer{
-		Bytes: make([]byte, 0),
-	}
+		// Cache not expired, return
+		return nil
+	} else if c.opts.CacheValidity > 0 {
+		// Use cached metadata for up to CacheValidity days.
+		// This is a bit of an hack, as we don't know when the
+		// last the it was updated, fallback to check the
+		// modification time of timestamp.json
+		if tm.Timestamp.Signed.IsExpired(time.Now()) {
+			// Always update if the timestamp is expired
+			return c.Refresh()
+		}
 
-	err = tufClient.Download(TrustedRootTUFPath, writer)
-	if err != nil {
-		return nil, err
-	}
+		var p = filepath.Join(
+			c.opts.CachePath,
+			URLToPath(c.opts.RepositoryBaseURL),
+			"timestamp.json",
+		)
+		fi, err := os.Stat(p)
+		if err != nil {
+			// Failed to get info on the file, fall back
+			// and update if needed
+			return c.Refresh()
+		}
 
-	err = fileJSONStore.SetMeta(TrustedRootTUFPath, writer.Bytes)
-	if err != nil {
-		return nil, err
+		if fi.ModTime().After(time.Now().Add(
+			time.Duration(-24*c.opts.CacheValidity) * time.Hour)) {
+			// No need to update
+			return nil
+		}
+		// A TUF client refresh will now happen (c.Refresh),
+		// update the mod time for the timestamp.
+		//
+		// Ignore the error here, there is no need to fail
+		// operation only because the file's metadata could
+		// not be updated
+		//nolint:errcheck
+		os.Chtimes(p, time.Now(), time.Now())
```

Comment:
Does this need to happen? Messing with the time of on-disk files feels a bit unexpected

Does `c.up.Refresh` forcefully update the on-disk metadata, as in downloads a new timestamp and writes it to the metadata folder regardless if the timestamp file has changed?

Reply:
No, go-tuf does not modify the disk of they downloaded timestamp is equal to the current one. So this is a hack to implement the non-standard metadata cache Sigstore want to have.

Reply:
What do you think about persisting this state in a separate file? We had something like this for the previous client, to persist the location of the remote repo. Might be helpful in the future if we add more configuration too.

Reply:
Good idea, relying a separate file is a better IMHO!

Reply:
Fixed now.